compiler/rustc_query_system/src/query/caches.rs

```diff
@@ -19,9 +19,9 @@ pub trait QueryCache: Sized {
     type Value: Copy;
 
     /// Checks if the query is already computed and in the cache.
-    fn lookup(&self, key: &Self::Key) -> Option<(Self::Value, DepNodeIndex)>;
+    fn lookup(&self, key: &Self::Key, key_hash: u64) -> Option<(Self::Value, DepNodeIndex)>;
```

Comment:
Is there a way to abstract over the `(Key, u64)` pair? A `PreHashedKey` struct to bundle both together?

Reply:
Maybe. I did measure passing around the key hash separately:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.7741s</td><td align="right">1.7677s</td><td align="right"> -0.36%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2775s</td><td align="right">0.2759s</td><td align="right"> -0.55%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.9941s</td><td align="right">0.9933s</td><td align="right"> -0.08%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.5865s</td><td align="right">1.5878s</td><td align="right"> 0.08%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">6.1135s</td><td align="right">6.1049s</td><td align="right"> -0.14%</td></tr><tr><td>Total</td><td align="right">10.7456s</td><td align="right">10.7297s</td><td align="right"> -0.15%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">0.9979s</td><td align="right"> -0.21%</td></tr></table>

It's a bit close to the noise floor, so I should rerun it with more samples. Most of the benefit of this PR seems to be the removal of the extra hashmap lookup.